src/librustc_typeck/check/coercion.rs

```diff
@@ -162,6 +163,26 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
         }
 
         if a.is_never() {
+            // Subtle: If we are coercing from `!` to `?T`, where `?T` is an unbound
+            // type variable, we *typically* want `?T` to fallback to `!` if not
+            // otherwise constrained. An example where this arises:
+            //
+            //     let _: Option<?T> = Some({ return; });
+            //
+            // here, we would coerce from `!` to `?T`. However, we do
+            // NOT want to do this in LUB mode, since we do not want
+            // this match to get the same treatment:
+            //
+            //     let _: Option<?T> = match x {
+            //         Ok(_) => Default::default(),
```

Comment:
The `_match` code uses `try_coerce` for the first arm, so you still have this problem if the arms are reversed.